### PR TITLE
Kubernetes support

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
This pull request introduces experimental support for running CML on Kubernetes clusters.

* **Supersedes** https://github.com/iterative/terraform-provider-iterative/pull/54 with per-cloud template golden tests. 😅 
* **Requires** https://github.com/iterative/cml/pull/403 and https://github.com/iterative/cml/pull/391 released.

### CODE REVIEW

**Comment by 0x2b3bfa0** [comment]
Note: https://github.com/iterative/terraform-provider-iterative/pull/53 would help uncluttering the onlline diff.

**Comment by 0x2b3bfa0** [comment]
Defenestrating pull request narrow scoping out of the highest window, but happy to offer a self-contained and verifiable set of changes. This pull request can be checked against #58, tested locally, and verified on the [`vgrengvir` parallel universe](https://github.com/vgrengvir/terraform-provider-iterative/runs/1918152376) organization.

**Comment by 0x2b3bfa0** [comment]
I could't resist the temptation of solving the only major remaining pain point that I was aware of: uniform resource identifiers. 🙃 
One more time, end-to-end tests have been [successful](https://github.com/vgrengvir/cml_kubernetes_case/runs/1923237393).

**Comment by DavidGOrtega** [comment]
Before merging this let's do the await based on logs.
What do you think?

**Comment by 0x2b3bfa0** [comment]
I would prefer to do it the other way around to avoid the [human] performance hit of rebasing this pull request over the drastic changes introduced by #65. Nevertheless, it's your call.

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, the basic GPU checks in AKS worked out of the box. :tada:

I would consider this pull request as completely tested and ready to merge. Once you approve it and we integrate the changes, I'll start working on a common approach for #65 and a GPU-specific version of the CML Docker container that serves as some sort of "Kubernetes Machine Image", so users don't need to install a couple of libraries as part of their workflow.

**Comment by DavidGOrtega** [comment]
@0x2b3bfa0 this PR has conflicts. Are you fixing them?
I want to merge it into contrib to do a final check.

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, are we merging this without #70? If true, please keep in mind that it depends on the iterative/cml#391 as-is.

🤖  Resolving conflicts...

**Comment by DavidGOrtega** [comment]
@0x2b3bfa0 understood! Lets merge #70 first. Im checking it.

File: iterative/aws/provider.go [hunk 1]
```diff
@@ -336,7 +339,25 @@ func awsClient(region string) (*ec2.EC2, error) {
...
+	"ap-southeast-2",
+	"ap-northeast-1",
+	"ca-central-1",
+	"eu-central-1",
+	"eu-west-1",
+	"eu-west-2",
+	"eu-west-3",
+	"eu-north-1",
+	"sa-east-1",
+}
```
Comment (0x2b3bfa0) [inline iterative/aws/provider.go:360]:
```suggestion
var ImageRegions = []string{"us-east-1", "us-east-2", "us-west-1", "us-west-2", "eu-north-1", "eu-west-1", "eu-west-2"}
```

File: README.md [hunk 2]
```diff
@@ -322,28 +488,29 @@ terraform {
...
+- Kubernetes
 
 <details>
 <summary>AWS instance equivalences</summary>
 <p>
 
-The instance type in AWS is calculated by joining the `instance_type` and
-`instance_gpu` values.
+The instance type in AWS is calculated joining the `instance_type` and
+`instance_gpu`
```
Comment (0x2b3bfa0) [inline README.md:513]:
```suggestion
The instance type in AWS is calculated by joining the `instance_type` and
`instance_gpu` values.
```

File: README.md [hunk 3]
```diff
@@ -396,11 +563,37 @@ The instance type in Azure is calculated by joining the `instance_type` and
...
+
+</p>
+</details>
+
+# `iterative-cml` image
 
-We've created a GPU-ready image based on Ubuntu 18.04. It comes with the
-following stack already installed:
+It's a GPU ready image based on Ubuntu 18.04. It has the following stack already
+installed:
```
Comment (0x2b3bfa0) [inline README.md:595]:
```suggestion
We've created a GPU-ready image based on Ubuntu 18.04. It comes with the
following stack already installed:
```

File: README.md [hunk 4]
```diff
@@ -396,11 +563,37 @@ The instance type in Azure is calculated by joining the `instance_type` and
...
 
-- Nvidia drivers
-- Docker
-- Nvidia-docker
+- nvidia drivers
+- docker
+- nvidia-docker
```
Comment (0x2b3bfa0) [inline README.md:599]:
```suggestion
- Nvidia drivers
- Docker
- Nvidia-docker
```

File: README.md [hunk 5]
```diff
@@ -396,7 +563,33 @@ The instance type in Azure is calculated by joining the `instance_type` and
...
+| xl   | tesla | 64        | 512 GiB |
+
+_Note: the resource limits specified are roughly equivalent to the ones from the
+equivalent AWS instances, but won't be allocated unless required by the running
+processes._
+
+</p>
+</details>
+
+# `iterative-cml` image
```
Comment (0x2b3bfa0) [inline README.md:592]:
```suggestion
## The iterative-cml image
```

File: README.md [hunk 6]
```diff
@@ -371,7 +538,7 @@ The instance type in AWS is calculated by joining the `instance_type` and
 <summary>Azure instance equivalences</summary>
 <p>
 
-The instance type in Azure is calculated by joining the `instance_type` and
+The instance type in Azure is calculated joining the `instance_type` and
```
Comment (0x2b3bfa0) [inline README.md:541]:
```suggestion
The instance type in Azure is calculated by joining the `instance_type` and
```

File: README.md [hunk 7]
```diff
@@ -322,21 +488,22 @@ terraform {
...
-  instance_type = "t2.micro"
+  instance_type     = "t2.micro"
 }
 ```
 
 </p>
 </details>
 
-## Supported vendors
+# Supported vendors
```
Comment (0x2b3bfa0) [inline README.md:502]:
```suggestion
## Supported vendors
```

File: README.md [hunk 8]
```diff
@@ -322,21 +488,22 @@ terraform {
...
 
-The Iterative Provider currently supports AWS and Azure. Google Cloud Platform
-is not currently supported.
+- AWS
+- Azure
+- Kubernetes
```
Comment (0x2b3bfa0) [inline README.md:506]:
```suggestion
The Iterative Provider currently supports AWS, Azure and Kubernetes. Google Cloud Platform
is not currently supported.
```

File: README.md [hunk 9]
```diff
@@ -252,61 +422,57 @@ terraform {
...
+| `ssh_private`       |                                          |                                                                           | SSH private in PEM format. If not provided one private and public key will be automatically generated and returned in terraform.tfstate                                                                                                                                                       |
 | `startup_script`    |                                          |                                                                           | Startup script also known as userData on AWS and customData in Azure. It can be expressed as multiline text using [TF heredoc syntax ](https://www.terraform.io/docs/configuration-0-11/variables.html)                                                                                       |
 
-## Requirements
+# Pitfalls
```
Comment (0x2b3bfa0) [inline README.md:466]:
```suggestion
## Requirements
```

File: README.md [hunk 10]
```diff
@@ -252,61 +422,57 @@ terraform {
...
-launch [instances from supported cloud vendors](#Supported-vendors). Please
-ensure that you have sufficient quotas with your cloud provider for the
-instances you intend to provision with Iterative Provider. If you're just
-starting out with a new account with a vendor, we recommend trying Iterative
-Provider with approved instances, such as the `t2.micro` instance for AWS.
+To be able to use the `instance_type` and `instance_gpu` you will need also to
+be allowed to launch [such instances](#Supported-vendors) within you cloud
+provider. Normally all the GPU instances need to be approved prior to be used by
+your vendor. You can always try with an already approved instance type by your
+vendor just setting it i.e. `t2.micro`
```
Comment (0x2b3bfa0) [inline README.md:472]:
```suggestion
To be able to use `instance_type` and `instance_gpu`, you'll need access to
launch [instances from supported cloud vendors](#Supported-vendors). Please
ensure that you have sufficient quotas with your cloud provider for the
instances you intend to provision with Iterative Provider. If you're just
starting out with a new account with a vendor, we recommend trying Iterative
Provider with approved instances, such as the `t2.micro` instance for AWS.
```

**Pull request merged**